File: sdk/go/sandbox.go
package liteboxd

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"time"
)

// SandboxService handles sandbox operations.
type SandboxService struct {
	client *Client
}

// Create creates a sandbox from a template.
//
// The template parameter is required. Use overrides to customize CPU, memory, TTL, or env.
// Overrides supported: cpu, memory, ttl, env
// NOT overridable: image, startupScript, files, readinessProbe (from template)
func (s *SandboxService) Create(ctx context.Context, template string, overrides *SandboxOverrides) (*Sandbox, error) {
	req := &CreateSandboxRequest{
		Template:  template,
		Overrides: overrides,
	}
	var result Sandbox
	err := s.client.doJSON(ctx, "POST", s.client.buildPath("sandboxes"), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// CreateWithVersion creates a sandbox from a specific template version.
func (s *SandboxService) CreateWithVersion(ctx context.Context, template string, version int, overrides *SandboxOverrides) (*Sandbox, error) {
	req := &CreateSandboxRequest{
		Template:        template,
		TemplateVersion: version,
		Overrides:       overrides,
	}
	var result Sandbox
	err := s.client.doJSON(ctx, "POST", s.client.buildPath("sandboxes"), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// List retrieves all sandboxes.
func (s *SandboxService) List(ctx context.Context) ([]Sandbox, error) {
	var result SandboxListResponse
	err := s.client.doJSON(ctx, "GET", s.client.buildPath("sandboxes"), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return result.Items, nil
}

// Get retrieves a specific sandbox by ID.
func (s *SandboxService) Get(ctx context.Context, id string) (*Sandbox, error) {
	var result Sandbox
	err := s.client.doJSON(ctx, "GET", s.client.buildPath("sandboxes", id), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Delete removes a sandbox.
func (s *SandboxService) Delete(ctx context.Context, id string) error {
	return s.client.doEmptyResponse(ctx, "DELETE", s.client.buildPath("sandboxes", id), nil, nil)
}

// Execute runs a command in the sandbox.
func (s *SandboxService) Execute(ctx context.Context, id string, command []string, timeout int) (*ExecResponse, error) {
	req := &ExecRequest{
		Command: command,
		Timeout: timeout,
	}
	var result ExecResponse
	err := s.client.doJSON(ctx, "POST", s.client.buildPath("sandboxes", id, "exec"), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// GetLogs retrieves container logs and Pod events.
func (s *SandboxService) GetLogs(ctx context.Context, id string) (*LogsResponse, error) {
	var result LogsResponse
	err := s.client.doJSON(ctx, "GET", s.client.buildPath("sandboxes", id, "logs"), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// UploadFile uploads a file to the sandbox.
func (s *SandboxService) UploadFile(ctx context.Context, id, filePath string, content []byte, contentType string) error {
	// Create multipart form body
	var body bytes.Buffer
	writer := multipart.NewWriter(&body)

	// Add file field
	part, err := writer.CreateFormFile("file", "file")
	if err != nil {
		return fmt.Errorf("failed to create form file: %w", err)
	}
	if _, err := part.Write(content); err != nil {
		return fmt.Errorf("failed to write file content: %w", err)
	}

	// Add path field
	if err := writer.WriteField("path", filePath); err != nil {
		return fmt.Errorf("failed to write path field: %w", err)
	}

	// Close writer to finalize the form
	if err := writer.Close(); err != nil {
		return fmt.Errorf("failed to close multipart writer: %w", err)
	}

	// Make request
	resp, err := s.client.doRequestWithReader(ctx, "POST", s.client.buildPath("sandboxes", id, "files"), &body, writer.FormDataContentType(), nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Check for error status
	if resp.StatusCode >= 400 {
		return handleErrorResponse(resp)
	}

	// Decode response
	var result map[string]string
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil
	}

	return nil
}

// DownloadFile downloads a file from the sandbox.
func (s *SandboxService) DownloadFile(ctx context.Context, id, filePath string) ([]byte, error) {
	queryParams := map[string]string{"path": filePath}

	resp, err := s.client.doRequest(ctx, "GET", s.client.buildPath("sandboxes", id, "files"), nil, queryParams)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Check for error status
	if resp.StatusCode >= 400 {
		return nil, handleErrorResponse(resp)
	}

	// Read binary content
	return io.ReadAll(resp.Body)
}

// WaitForReady waits until the sandbox reaches running status.
// pollInterval is the time between checks (default 2s).
// timeout is the maximum wait time (default 5m).
func (s *SandboxService) WaitForReady(ctx context.Context, id string, pollInterval, timeout time.Duration) (*Sandbox, error) {
	if pollInterval == 0 {
		pollInterval = 2 * time.Second
	}
	if timeout == 0 {
		timeout = 5 * time.Minute
	}

	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	ticker := time.NewTicker(pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("timeout waiting for sandbox to be ready")
		case <-ticker.C:
			sandbox, err := s.Get(ctx, id)
			if err != nil {
				return nil, err
			}
			if sandbox.Status == SandboxStatusRunning {
				return sandbox, nil
			}
			if sandbox.Status == SandboxStatusFailed {
				return nil, fmt.Errorf("sandbox failed")
			}
		}
	}
}

File: sdk/python/src/liteboxd/services/sandbox.py
"""Sandbox service for LiteBoxd SDK."""

from __future__ import annotations

import builtins
import time
from typing import TYPE_CHECKING, Any

from ..exceptions import SandboxFailedError
from ..exceptions import TimeoutError as LiteBoxdTimeoutError
from ..models import (
    CreateSandboxRequest,
    ExecResult,
    LogsResult,
    Sandbox,
    SandboxOverrides,
    SandboxStatus,
    to_api_dict,
)

if TYPE_CHECKING:
    from ..client import Client


class SandboxService:
    """Sandbox operations service."""

    def __init__(self, client: "Client") -> None:
        """Initialize sandbox service."""
        self._client = client

    def create(
        self,
        template: str,
        *,
        template_version: int | None = None,
        overrides: SandboxOverrides | None = None,
    ) -> Sandbox:
        """
        Create a sandbox from a template.

        Args:
            template: Template name (required)
            template_version: Template version (optional, defaults to latest)
            overrides: Configuration overrides (cpu, memory, ttl, env)

        Returns:
            Created sandbox object

        Raises:
            NotFoundError: Template not found
            BadRequestError: Invalid request parameters
        """
        request = CreateSandboxRequest(
            template=template,
            template_version=template_version,
            overrides=overrides,
        )
        response = self._client._post("sandboxes", json=to_api_dict(request))
        return Sandbox.model_validate(response.json())

    def list(self) -> builtins.list[Sandbox]:
        """
        List all sandboxes.

        Returns:
            List of sandbox objects
        """
        response = self._client._get("sandboxes")
        data = response.json()
        return [Sandbox.model_validate(item) for item in data.get("items", [])]

    def get(self, sandbox_id: str) -> Sandbox:
        """
        Get a specific sandbox by ID.

        Args:
            sandbox_id: Sandbox ID

        Returns:
            Sandbox object

        Raises:
            NotFoundError: Sandbox not found
        """
        response = self._client._get(f"sandboxes/{sandbox_id}")
        return Sandbox.model_validate(response.json())

    def delete(self, sandbox_id: str) -> None:
        """
        Delete a sandbox.

        Args:
            sandbox_id: Sandbox ID
        """
        self._client._delete(f"sandboxes/{sandbox_id}")

    def execute(
        self,
        sandbox_id: str,
        command: builtins.list[str],
        *,
        timeout: int = 30,
    ) -> ExecResult:
        """
        Execute a command in the sandbox.

        Args:
            sandbox_id: Sandbox ID
            command: Command to execute as list of strings
            timeout: Execution timeout in seconds

        Returns:
            Command execution result (exit_code, stdout, stderr)

        Raises:
            NotFoundError: Sandbox not found
        """
        payload: dict[str, Any] = {
            "command": command,
            "timeout": timeout,
        }
        response = self._client._post(f"sandboxes/{sandbox_id}/exec", json=payload)
        return ExecResult.model_validate(response.json())

    def get_logs(self, sandbox_id: str) -> LogsResult:
        """
        Get sandbox logs and events.

        Args:
            sandbox_id: Sandbox ID

        Returns:
            Logs and events
        """
        response = self._client._get(f"sandboxes/{sandbox_id}/logs")
        return LogsResult.model_validate(response.json())

    def upload_file(
        self,
        sandbox_id: str,
        path: str,
        content: bytes,
        *,
        content_type: str = "application/octet-stream",
    ) -> None:
        """
        Upload a file to the sandbox.

        Args:
            sandbox_id: Sandbox ID
            path: Target path inside the container
            content: File content as bytes
            content_type: Content type (default: application/octet-stream)
        """
        files = {
            "file": ("file", content, content_type),
        }
        data = {
            "path": path,
        }
        self._client._post(f"sandboxes/{sandbox_id}/files", data=data, files=files)

    def download_file(self, sandbox_id: str, path: str) -> bytes:
        """
        Download a file from the sandbox.

        Args:
            sandbox_id: Sandbox ID
            path: File path inside the container

        Returns:
            File content as bytes
        """
        response = self._client._get(
            f"sandboxes/{sandbox_id}/files",
            params={"path": path},
        )
        return response.content

    def wait_for_ready(
        self,
        sandbox_id: str,
        *,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
    ) -> Sandbox:
        """
        Wait until the sandbox reaches running status.

        Args:
            sandbox_id: Sandbox ID
            poll_interval: Time between checks in seconds (default: 2.0)
            timeout: Maximum wait time in seconds (default: 300.0)

        Returns:
            Sandbox object in running state

        Raises:
            TimeoutError: Wait timed out
            SandboxFailedError: Sandbox failed to start
        """
        start_time = time.monotonic()
        deadline = start_time + timeout

        while True:
            sandbox = self.get(sandbox_id)

            if sandbox.status == SandboxStatus.RUNNING:
                return sandbox

            if sandbox.status == SandboxStatus.FAILED:
                raise SandboxFailedError(sandbox_id, "Sandbox failed to start")

            if sandbox.status == SandboxStatus.SUCCEEDED:
                raise SandboxFailedError(sandbox_id, "Sandbox exited unexpectedly")

            # Check timeout
            if time.monotonic() >= deadline:
                raise LiteBoxdTimeoutError(
                    f"Timeout waiting for sandbox {sandbox_id} to be ready "
                    f"(current status: {sandbox.status})"
                )

            # Wait before next poll
            time.sleep(poll_interval)
